src/java/org/apache/cassandra/triggers/TriggerExecutor.java

```diff
@@ -220,6 +230,16 @@ private List<Mutation> executeInternal(PartitionUpdate update)
         Triggers triggers = update.metadata().triggers;
         if (triggers.isEmpty())
             return null;
+        Config.TriggersPolicy policy = DatabaseDescriptor.getTriggersPolicy();
```

Comment:
I just feel like we are burying this logic in some executeInternal method instead of failing as the very first thing we have a chance to do that.

Reply:
> at least DatabaseDescriptor.getTriggersPolicy() might be "cached" so we do not need to call it all over again.

As the policy is effectively immutable, this is almost equivalent to a constant lookup. Caching it isn't going to give us much in performance terms, and it won't do anything for code clarity. If it were memoized on `TriggersExecutor`, tests would need a way to set/override it, so for the sake of clarity I'd suggest leaving it as is. 

> I just feel like we are burying this logic in some executeInternal method instead of failing as the very first thing we have a chance to do that.

This is the first point at which we inspect the update to check whether the update actually has any triggers attached, so it seems a reasonable place to do this validation. It's intrinsic to the execution of any triggers, so I'd say the method that encapsulates that (i.e. `executeInternal`) is the ideal place for it.

> I would also move it to execute method, outside of the for loop where executeInternal is executed, because we would unnecessarily go through this logic all over again calling warn on logger (even it is non-spamming one).

Triggers are associated with a specific `PartitionUpdate` so in the version of `execute` which takes a collection, this has to be checked for each mutation.

Reply:
yeah it would be a little bit hairy the other way ... we also do not have access to `triggers` from `update.metadata().triggers` unless we loop over `PartitionUpdate`s for the sake of using them in logger or exception upon message construction. 

Just to be explicit, so this would throw an exception on the first `Mutation` and on the first `PartitionUpdate` of that mutation.

Reply:
It would throw on the first `PartitionUpdate` which has a trigger. It may iterate over other updates which don't before it gets to one, but that's fine too.